File: src/main/java/liqp/tags/Include.java
package liqp.tags;

import java.io.File;
import liqp.Template;
import liqp.nodes.LNode;
import liqp.nodes.RenderContext;

public class Include extends Tag {

  public static final String INCLUDES_DIRECTORY_KEY = "liqp@includes_directory";
  public static String DEFAULT_EXTENSION = ".liquid";

  @Override
  public Object render(RenderContext context, LNode... nodes) {

    boolean entered = false;
    try {
      String includeResource = super.asString(nodes[0].render(context));
      String extension = DEFAULT_EXTENSION;
      if (includeResource.indexOf('.') > 0) {
        extension = "";
      }
      File includeResourceFile;
      File includesDirectory = (File) context.get(INCLUDES_DIRECTORY_KEY);
      if (includesDirectory != null) {
        includeResourceFile = new File(includesDirectory, includeResource + extension);
      } else {
        includeResourceFile = new File(context.getTemplateFactory().getFlavor().snippetsFolderName, includeResource +
              extension);
      }
      Template template = context.getTemplateFactory().parseFile(includeResourceFile);

      // check if there's a optional "with expression"
      if (nodes.length > 1) {
        Object value = nodes[1].render(context);
        context.set(includeResource, value);
      }

      context.addFrame();
      entered = true;

      return context.render(template);
    } catch (Exception e) {
      return "";
    } finally {
      if (entered) {
        context.popFrame();
      }
    }
  }
}

File: liqp-core/src/main/kotlin/liqp/tags/Include.kt
package liqp.tags

import liqp.LValue
import liqp.nodes.LNode
import liqp.nodes.RenderContext

const val DEFAULT_EXTENSION = ".liquid"

class Include : Tag() {

  override fun render(context: RenderContext, vararg nodes: LNode): Any {

    var entered = false
    try {
      val includeResource = LValue.asString(nodes[0].render(context))
      val extension = when {
        '.' in includeResource -> ""
        else -> DEFAULT_EXTENSION
      }

      val includesDir = context.includesDir
      val includeResourceFile = includesDir.resolve(includeResource + extension)

      // This will take advantage of caching for better performance, but we should probably
      // also write the template locally
      val template = context.parser.parseFile(includeResourceFile)

      // Push the frame prior to setting the variable.  This ensures that the variable doesnt
      // leak to any siblings
      context.pushFrame()
      entered = true

      // check if there's a optional "with expression"  If so, push
      if (nodes.size > 1) {
        val value = nodes[1].render(context)
        context[includeResource] = value
      }


      return context.render(template)
    } catch (e: Exception) {
      context.logs += e.toString()
      return ""
    } finally {
      if (entered) {
        context.popFrame()
      }
    }
  }
}
